Sources/FrontEnd/Parse/Parser.swift

```diff
@@ -2960,6 +2961,180 @@ public enum Parser {
+        try skipConditionalCompilationBranch(in: &state, stoppingAtElse: false)
+        fallback = []
+      } else {
+        fallback = try parseConditionalCompilationBranch(in: &state)
+      }
+      // Expect #endif.
+      if state.take(.poundEndif) == nil {
+        throw [.error(expected: "#endif", at: state.currentLocation)] as DiagnosticSet
+      }
+    } else if let head2 = state.take(.poundElseif) {
+      if condition.mayNotNeedParsing && condition.holds(for: CompilerInfo.instance) {
+        try skipConditionalCompilationBranch(in: &state, stoppingAtElse: false)
+        fallback = []
+      } else {
+        // We continue with another conditional compilation statement.
+        fallback = [try parseCompilerConditionTail(head: head2, in: &state)]
+      }
+    } else {
+      throw [.error(expected: "statement, #endif, #else or #elseif", at: state.currentLocation)]
+        as DiagnosticSet
+    }
+
+    let r = state.insert(
+      ConditionalCompilationStmt(
+        condition: condition,
+        stmts: stmts,
+        fallback: fallback,
+        site: head.site.extended(upTo: state.currentIndex)))
+    return AnyStmtID(r)
+  }
+  /// Skips the branch body of a conditional compilation statement.
+  /// Skip over nested conditional compilation statements.
+  private static func skipConditionalCompilationBranch(
+    in state: inout ParserState, stoppingAtElse: Bool
+  ) throws {
+    var innerCompilerConditions = 0
+    while let head = state.peek() {
+      let closing: Bool
+      switch head.kind {
+      case .poundEndif: closing = true
+      case .poundElse, .poundElseif:
+        // Check if we need to stop at the corresponding #else[if].
+        closing = stoppingAtElse && innerCompilerConditions == 0
+      default: closing = false
+      }
+      if closing {
+        if innerCompilerConditions == 0 {
+          // We are done skipping all the tokens from the body.
+          break
+        } else {
+          innerCompilerConditions -= 1
+        }
+      }
+      // Consume this token.
+      _ = state.take()
+
+      // Are opening another nested compiler condition?
+      if head.kind == .poundIf {
+        innerCompilerConditions += 1
+      }
+    }
+  }
+  /// Parses the body of a compiler condition structure branch.
+  private static func parseConditionalCompilationBranch(in state: inout ParserState) throws
+    -> [AnyStmtID]
+  {
+    let stmtsParser =
+      (zeroOrMany(take(.semi))
+        .and(zeroOrMany(stmt.and(zeroOrMany(take(.semi))).first))
+        .second)
+    return try stmtsParser.parse(&state) ?? []
+  }
+
+  /// Parses the condition of a compiler conditional.
+  private static func parseCompilerCondition(in state: inout ParserState) throws
+    -> ConditionalCompilationStmt.Condition
+  {
+    if let boolLiteral = state.take(.bool) {
+      return state.lexer.sourceCode[boolLiteral.site] == "true" ? .`true` : .`false`
+    } else if let name = state.take(.name) {
```

Comment:
```suggestion
    }
    
    if let name = state.take(.name) {
```
No need for a `else if` If the first branch returns.

Reply:
If it's ok with you, I would prefer to keep it like this.
It's easier to read, even without looking at the body of different branches.
Moreover, the compiler will warn you if you forget a `return`.

Reply:
I disagree and would really prefer to avoid writing unnecessary `else if` clauses. I think that having separate blocks of code helps readability and this style is more consistent with the rest of our code.

Reply:
Fixed.
FYI: https://www.youtube.com/watch?v=SFv8Wm2HdNM